src/utils/blockedMonitor.ts

```diff
@@ -0,0 +1,32 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License.
+
+import { AzureFunctionsRpcMessages as rpc } from './../../azure-functions-language-worker-protobuf/src/rpc';
+import LogCategory = rpc.RpcLog.RpcLogCategory;
+import LogLevel = rpc.RpcLog.Level;
+import blockedAt = require('blocked-at');
+
+export function startBlockedMonitor(channel: { log: (log: rpc.IRpcLog) => void }, threshold = 100): NodeJS.Timer {
+    let blockedHistory: { time: string; duration: number; stack: string; resource: string }[] = [];
+
+    //threshold - minimum miliseconds of blockage to report.
+    //other parameters are default, more details on https://github.com/naugtur/blocked-at.
+    blockedAt(
+        (ms, stack, resource) => {
+            blockedHistory.push({ time: new Date().toUTCString(), duration: ms, stack: stack, resource: resource });
+        },
+        { threshold: threshold }
+    );
+
+    // Log blockedHistory if it's not empty each 10 seconds
+    return setInterval(() => {
```

Comment:
Here is what the function would look like if you combine all my comments (please read those first).

IMO the main unknowns are (1) if you're okay doing a threshold of 5000 as that impacts a lot of the code structure and (2) if you still want to print `resource` depending on the answer to my question about it

```typescript
export function startBlockedMonitor(channel: { log: (log: rpc.IRpcLog) => void }, threshold = 5000): void {
    function systemWarn(message: string) {
        channel.log({
            message,
            level: LogLevel.Warning,
            logCategory: LogCategory.System,
        });
    }

    systemWarn(
        `Warning: Monitoring for blocking code is turned on, with a threshold of ${threshold} ms. This will have a negative impact on performance. Adjust "AZURE_FUNCTIONS_NODE_BLOCK_LOG" to turn it off.`
    );

    //threshold - minimum milliseconds of blockage to report.
    //other parameters are default, more details on https://github.com/naugtur/blocked-at.
    blockedAt(
        (ms, stack, _resource) => {
            systemWarn(`Warning: Event loop blocked for ${Math.round(ms)} ms.`);
            systemWarn(
                `IMPORTANT NOTE: The following stack trace is only an approximation and you should analyze all synchronous operations from that point onward:`
            );
            systemWarn(stack.join(EOL));
        },
        { threshold: threshold }
    );
}
```

Reply:
5 secs threshold looks long for me - it would report only delays > 5 secs.
My idea was to detect delays < 100ms, accumulate history and report each 10 seconds so the grpc channel would be overwhelmed with the logs. In case we are using `systemWarn`(console logs) I am not sure about perf impact if we have to many logs.

Regarding `resource` - if the signature provides the parameter why do not log it even it can be not very useful?